Fix TIDAL support as with given API token it doesn't use rtmp

theli-ua opened the pull request:
Tested with LOSSLESS setting - works for me

tamland commented:
Thanks for taking the time for this! just have a few point that can be improved;) Can you also submit those to https://github.com/tamland/wimpy ? Trying to keep a separate library usable outside of kodi

theli-ua commented:
There must be more efficient way to manage library outside. git submodule maybe?

theli-ua commented:
I've addressed your comments and added support for browsing Genres
I'll submit pr for wimpy changes once you accept this one so that I won't need to keep them both in sync while doing changes if you don't mind

tamland commented on addon.py:
```diff
@@ -231,9 +310,13 @@ def logout():
     media_url = wimp.get_media_url(track_id)
-    host, app, playpath = media_url.split('/', 3)
-    rtmp_url = 'rtmp://%s app=%s playpath=%s' % (host, app, playpath)
+    if media_url.startswith('http://') and not  media_url.startswith('https://'):
```
This logic isn't correct

tamland commented on addon.py:
```diff
@@ -89,10 +92,86 @@ def urls_from_id(view_func, items):
+    add_directory('Tracks', genre_tracks, _genre = _genre)
+    xbmcplugin.endOfDirectory(plugin.handle)
+
+@plugin.route('/genre/<_genre>/playlists')
```
dont think those underscores are necessary

tamland commented on lib/wimpy/wimpy.py:
```diff
@@ -50,6 +50,7 @@ class Config(object):
     quality = Quality.high
     session_id = None
     user_id = None
+    mimetype = ''
```
leftover

tamland commented on addon.py:
```diff
@@ -89,10 +92,86 @@ def urls_from_id(view_func, items):
+
+@plugin.route('/whats_new')
+def whats_new():
+    add_directory('New Playlists', featured_playlists, _type = 'new')
```
follow existing format (no spaces in keywords). should be`_type='new'`

tamland commented on addon.py:
```diff
@@ -89,10 +92,86 @@ def urls_from_id(view_func, items):
+def featured_tracks(_type):
+    items = wimp.get_recommended_new_top('tracks', _type)
+    track_list(items)
+@plugin.route('/featured/playlists/<_type>')
```
should two spaces between functions

tamland commented:
Sorry for taking so long to get back. Can you please take the time to follow the existing pep8 style? (I've added a few comments)

And please squash changes into the commit it belongs.

tamland commented:
>There must be more efficient way to manage library outside. git submodule maybe?

submodule doesn't' cut it as it's folder for a different project, not the entire project. And it should work with zip download. I just mount the `wimpy` folder in `lib` locally. Alternatively you can set up subtree-merging if you want

theli-ua commented:
Not really following "squash" part. Do you want me to squash everything into a single commit?

tamland commented:
Not into a single commit. There are ~6 different features added, the there should 6 distinct commits that doesn't change other random pars. If that isn't too hard or impossible to do now. It's really a lot easier to use git rebase+edit from the beginning.

theli-ua committed:
PEP8-ify

theli-ua committed:
Fix TIDAL support

theli-ua committed:
Add featured playlists aka promotions

theli-ua committed:
Add What's New aka recommended/new/top albums/playlists/tracks

theli-ua committed:
Add 'moods' support

theli-ua committed:
Add support for genre browsing

theli-ua committed:
Add 'track radio' context menu item for tracks

theli-ua commented:
>If that isn't too hard or impossible to do now.
With git's interactive patch staging its not that hard. I did what you asked for.

theli-ua commented:
Also, when/if you accept this I'll post same changeset for wimpy in a separate PR. Or do you want that one in a single commit?

tamland commented:
Ok, cherry-picked into master branch. Squashed one into my own, and I dropped the 'track radio' one. It doesn't work for me. I get 404. Want to double check that?

Same changeset for wimpy please. In the future you can send changes to wimpy to that repo, and plugin changes here. Then I can do one big sync commit to this repo. It's much easier to manage that way.

theli-ua commented:
Track Radio works for me perfectly fine. Could you take a look what xhr request does wimp make from web player if you try to go to track radio?

tamland commented:
You forgot the `limit` parameter. Default it 9999 so it just gives 'Track radio cannot be generated' here. Fixed with 2c9a1421a34fbba479bc1f571200efcefd7cd741

theli-ua commented:
It just works for me without limiting to less then 9999 on TIDAL:

    "limit": 9999,
    "offset": 0,
    "totalNumberOfItems": 708
